<a href="https://colab.research.google.com/github/sskamalapurkar/Unsupervised_ML_Clustering/blob/main/Online_Retail_Customer_Segmentation_Sayali_Kamalapurkar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Problem Statement**

---



In this project, your task is to identify major customer segments on a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

**In this project, we are required to :**


*   Understand the Dataset & cleanup (if required)

*   Build a clustering model to segment the customer-based similarity

*   Also fine-tune the hyperparameters & compare the evaluation metrics of various classification algorithms

# **Introduction**

---



In [1]:
#importing the libraries
import pandas as pd
import numpy as np
from numpy import math
from datetime import datetime

#Visualization libraries
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline


#ignore warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
# importing dataset
file_path = '/content/drive/MyDrive/Unsupervised ML - Clustering/Online Retail.xlsx'
DataFrame = pd.read_excel(file_path)
df = DataFrame.copy()

In [3]:
# let us have a glance at the dataset
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

# **Data Description**

---

***Attribute Information***

>|Field Name|Description|
|---|---|
|**InvoiceNo:**|Invoice number. A 6-digit integer number uniquely assigned to each transaction.
||If this code starts with letter 'c', it indicates a cancellation.|
|**StockCode:**|Product (item) code. A 5-digit integer number uniquely assigned to each distinct product.|
|**Description:**|Product (item) name|
|**Quantity:**|The quantities of each product (item) per transaction.|
|**InvoiceDate:**|Invice Date and time. The day and time when each transaction was generated.|
|**UnitPrice:**|Unit price. Product price per unit in sterling.|
|**CustomerID:**|Customer number. A 5-digit integral number uniquely assigned to each customer.|
|**Country:**|Country name. The name of the country where each customer resides.|

In [4]:
#Checking number of rows and columns
df.shape

(541909, 8)

In [5]:
print("Total retail transactions including each product in every transactions (Rows): ", df.shape[0])
print("NUmber of features (Columns): ", df.shape[1])

Total retail transactions including each product in every transactions (Rows):  541909
NUmber of features (Columns):  8


In [6]:
# printing names of all the columns / features
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [7]:
# getting info about the columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


**There are 4 features whose datatype is object, 2 features whose datatype is float64, 1 feature whose datatype is datetime64 and 1 feature whose datatype is int64.**

**Memory usage by the dataset is 33.1 MB.**

In [8]:
#checking for unique values
df.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

**In the dataset, we have retail transactions from almost 38 different countries of the world.**

**Almost 25900 overall transactions, this includes cancelled transactions too.**

**Also, 4223 different items are listed in the dataset.**

**Overall customer count in the available dataset is 4372.**

In [9]:
#let us check for null values
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

**Only 2 columns have null values in it.**

# **Duplicate Values**

---

Why is it important to remove duplicate records from my data?

> "Duplication" means that we have repeated data in our dataset. This could be due to things like data entry errors or data collection methods. By removing duplicate values in our data set, time and money are saved by not sending identical communications multiple times to the same person.

In [10]:
# checking for duplicate Values in the dataset
duplicates=len(df[df.duplicated()])
print("The number of duplicate values in the dataset is - ",duplicates)

The number of duplicate values in the dataset is -  5268


In [11]:
df[df.duplicated()].head()

InvoiceNo StockCode                        Description  Quantity  \
517    536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527    536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537    536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539    536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555    536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517 2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
527 2010-12-01 11:45:00       2.10     17908.0  United Kingdom  
537 2010-12-01 11:45:00       2.95     17908.0  United Kingdom  
539 2010-12-01 11:45:00       4.95     17908.0  United Kingdom  
555 2010-12-01 11:49:00       2.95     17920.0  United Kingdom

In [12]:
# Select transactions with InvoiceNo = '536409' & StockCode = '21866'
df.loc[(df['InvoiceNo'].astype(str) == '536409') & (df['StockCode'].astype(str) == '21866')]

InvoiceNo StockCode                  Description  Quantity  \
494    536409     21866  UNION JACK FLAG LUGGAGE TAG         1   
517    536409     21866  UNION JACK FLAG LUGGAGE TAG         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
494 2010-12-01 11:45:00       1.25     17908.0  United Kingdom  
517 2010-12-01 11:45:00       1.25     17908.0  United Kingdom

As we can see, the duplicates are nothing but repeat entry of the particular transactions. These entries are exactly same, means there is no significance of such entries while analyzing the data. Hence we will drop all those duplicate rows from the data.

In [13]:
# Remove duplicates
df = df.drop(index=df[df.duplicated()].index)
print("Number of rows after removal of duplicates : {}".format(df.shape[0]))

Number of rows after removal of duplicates : 536641


In [14]:
df.describe(include = 'all')

InvoiceNo StockCode                         Description  \
count    536641.0    536641                              535187   
unique    25900.0      4070                                4223   
top      573585.0    85123A  WHITE HANGING HEART T-LIGHT HOLDER   
freq       1114.0      2301                                2357   
first         NaN       NaN                                 NaN   
last          NaN       NaN                                 NaN   
mean          NaN       NaN                                 NaN   
std           NaN       NaN                                 NaN   
min           NaN       NaN                                 NaN   
25%           NaN       NaN                                 NaN   
50%           NaN       NaN                                 NaN   
75%           NaN       NaN                                 NaN   
max           NaN       NaN                                 NaN   

             Quantity          InvoiceDate      UnitPrice     CustomerID  \
count   536641.000000               536641  536641.000000  401604.000000   
unique            NaN                23260            NaN            NaN   
top               NaN  2011-10-31 14:41:00            NaN            NaN   
freq              NaN                 1114            NaN            NaN   
first             NaN  2010-12-01 08:26:00            NaN            NaN   
last              NaN  2011-12-09 12:50:00            NaN            NaN   
mean         9.620029                  NaN       4.632656   15281.160818   
std        219.130156                  NaN      97.233118    1714.006089   
min     -80995.000000                  NaN  -11062.060000   12346.000000   
25%          1.000000                  NaN       1.250000   13939.000000   
50%          3.000000                  NaN       2.080000   15145.000000   
75%         10.000000                  NaN       4.130000   16784.000000   
max      80995.000000                  NaN   38970.000000   18287.000000   

               Country  
count           536641  
unique              38  
top     United Kingdom  
freq            490300  
first              NaN  
last               NaN  
mean               NaN  
std                NaN  
min                NaN  
25%                NaN  
50%                NaN  
75%                NaN  
max                NaN

**Observations:**



*   There are 4070 unique product codes. **'85123A'** is the most common product code appeared 2301 times.

*   Majority of customers resides in United Kingdom.


*   List item
*   List item





In [15]:
df.loc[(df['InvoiceNo'].astype(str) == '573585')]

InvoiceNo     StockCode                         Description  Quantity  \
430237    573585         11001         ASSTD DESIGN RACING CAR PEN         2   
430238    573585         15036           ASSORTED COLOURS SILK FAN        15   
430239    573585         15039                      SANDALWOOD FAN        15   
430240    573585       15056BL             EDWARDIAN PARASOL BLACK         1   
430241    573585        15056N           EDWARDIAN PARASOL NATURAL         1   
...          ...           ...                                 ...       ...   
431346    573585        90209C         PINK ENAMEL+GLASS HAIR COMB         3   
431347    573585        90214S           LETTER "S" BLING KEY RING         1   
431348    573585           DOT                      DOTCOM POSTAGE         1   
431349    573585  gift_0001_20  Dotcomgiftshop Gift Voucher £20.00         1   
431350    573585         23539             WALL ART LOVES' SECRET          2   

               InvoiceDate  UnitPrice  CustomerID         Country  
430237 2011-10-31 14:41:00       3.29         NaN  United Kingdom  
430238 2011-10-31 14:41:00       1.63         NaN  United Kingdom  
430239 2011-10-31 14:41:00       1.63         NaN  United Kingdom  
430240 2011-10-31 14:41:00      12.46         NaN  United Kingdom  
430241 2011-10-31 14:41:00      12.46         NaN  United Kingdom  
...                    ...        ...         ...             ...  
431346 2011-10-31 14:41:00       2.07         NaN  United Kingdom  
431347 2011-10-31 14:41:00       0.83         NaN  United Kingdom  
431348 2011-10-31 14:41:00    2019.05         NaN  United Kingdom  
431349 2011-10-31 14:41:00      16.67         NaN  United Kingdom  
431350 2011-10-31 14:41:00       5.79         NaN  United Kingdom  

[1114 rows x 8 columns]